File: packages/guards/src/schema-drift.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface SchemaDriftOptions {
  action: 'block' | 'warn';
  /** Expected top-level keys */
  expectedKeys: string[];
  /** Allow extra keys (default false) */
  allowExtra?: boolean;
}

export function schemaDrift(options: SchemaDriftOptions): Guard {
  const allowExtra = options.allowExtra ?? false;
  const expected = new Set(options.expectedKeys);

  return {
    name: 'schema-drift',
    version: '0.1.0',
    description: 'Detects schema drift in structured JSON output',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      let parsed: Record<string, unknown>;
      try { parsed = JSON.parse(text.trim()); } catch {
        return { guardName: 'schema-drift', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      }
      if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
        return { guardName: 'schema-drift', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      }
      const actual = new Set(Object.keys(parsed));
      const missing = [...expected].filter((k) => !actual.has(k));
      const extra = allowExtra ? [] : [...actual].filter((k) => !expected.has(k));
      const triggered = missing.length > 0 || extra.length > 0;
      return { guardName: 'schema-drift', passed: !triggered, action: triggered ? options.action : 'allow', latencyMs: Math.round(performance.now() - start), details: triggered ? { missing, extra } : undefined };
    },
  };
}

File: python/open_guardrail/guards/schema_drift.py
"""Detect JSON schema drift from expected keys."""

import json
import time
from typing import List

from open_guardrail.core import GuardResult


class _SchemaDrift:
    def __init__(
        self,
        *,
        action: str = "warn",
        expected_keys: List[str] | None = None,
        allow_extra: bool = False,
    ) -> None:
        self.name = "schema-drift"
        self.action = action
        self.expected_keys = set(expected_keys or [])
        self.allow_extra = allow_extra

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()

        try:
            data = json.loads(text)
        except (json.JSONDecodeError, TypeError):
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="schema-drift",
                passed=False,
                action=self.action,
                message="Invalid JSON",
                latency_ms=round(elapsed, 2),
            )

        if not isinstance(data, dict):
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="schema-drift",
                passed=False,
                action=self.action,
                message="JSON root is not an object",
                latency_ms=round(elapsed, 2),
            )

        actual = set(data.keys())
        missing = self.expected_keys - actual
        extra = actual - self.expected_keys

        triggered = bool(missing)
        if not self.allow_extra and extra:
            triggered = True

        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="schema-drift",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Schema drift detected"
                if triggered else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "missing": sorted(missing),
                "extra": sorted(extra),
            } if triggered else None,
        )


def schema_drift(
    *,
    action: str = "warn",
    expected_keys: List[str] | None = None,
    allow_extra: bool = False,
) -> _SchemaDrift:
    return _SchemaDrift(
        action=action,
        expected_keys=expected_keys,
        allow_extra=allow_extra,
    )
